server/static/dashboard/collab.js

```diff
@@ -7,42 +10,486 @@ collab.fetch(function(err) {
+      // Save data for regex updating
+      // TODO: Better way to do this?
+      $(diff).data('baseline', baseline);
+      $(diff).data('text', file.data.text); // TODO: update this data on update
+      $(diff).data('filepath', file.data.filepath);
+
+      var extraArgsForFile = Object.assign({'filepath': file.data.filepath}, extraArgs);
+
       if (cutoff) {
         $.ajax('/historical/' + project + '/' + collabid + '/' + file.data.filepath + '/' + cutoff).done(function(historical) {
-          updateDiff(diff, baseline, historical.data ? historical.data.text : undefined);
+
+          // Used for regex updating
+          if (historical.data) {
+            $(diff).data('text', historical.data.text); // TODO: update this data on update
+          }
+
+          updateFunction(diff, baseline, historical.data ? historical.data.text : undefined, extraArgsForFile);
+
         }).fail(function(req, status, err) {
           diff.textContent = 'Error fetching code: ' + errorToString(req.responseJSON, status, err);
         });
       } else {
         file.subscribe(function() {
-          updateDiff(diff, baseline, file.data.text);
-          file.on('op', function() {
-            updateDiff(diff, baseline, file.data.text);
+          updateFunction(diff, baseline, file.data.text, extraArgsForFile);
+          file.on('op', function(op) {
+            extraArgs["op"] = op;
+            updateFunction(diff, baseline, file.data.text, extraArgsForFile);
           });
         });
       }
     }).fail(function(req, status, err) {
       diff.textContent = 'Error fetching baseline: ' + errorToString(req.responseJSON, status, err);
     });
+
   });
-});
+}
 
-function updateDiff(node, baseline, text) {
+
+
+/////////////////////////////////////////////
+///////////// UPDATE FUNCTIONS //////////////
+// Displays DOM differently for different visualizations.
+
+function updateFunction(node, baseline, text, extraArgs) {
   if (baseline === undefined || text === undefined) { return; }
   node.innerHTML = '';
+  // TODO: This causes page to seem as if it's refreshing every
+  //   time you change the regex, and causes you to lose whatever
+  //   spot in the files you were at
+
+  var threshold = parameters['threshold'];
+  var regexes = parameters['regexes'];
+  var hideCommon = parameters['hideCommonPrefixes'];
+
+  if (!threshold) {
+    // No treshold means no total diff
+    drawNormalDiff(baseline, text, node);
+
+    if (regexes) {
+      addRegexHighlighting(node, regexes);
+      // TODO: Without regex matching, the entire line is highlighted
+      //   in light yellow if the student added that code.
+      // However, now only the part of the line with code on it is
+      //   highlighted.
+    }
+
+  } else {
+    // Get the flattened diff to display in total diff style
+    var filepath = extraArgs["filepath"];
+    var url = getAjaxUrlForTotalDiff(filepath, threshold);
```

Comment:
Can we get e.g. jQuery to construct this query string from an object and eliminate this helper function?

Reply:
Eliminated function because I eliminated threshold, which made this function much simpler.
Side note: You can do jquery to do this: http://api.jquery.com/jquery.param/